### PR TITLE
Query as entity

### PR DESCRIPTION
**Pull request by mohityadav766**
### Describe your changes :

- Added Query as a Sub Entity

#
### Type of change :

- [x] Bug fix
- [x] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      "$ref": "../../type/entityReference.json#/definitions/entityReferenceList"
+    },
+    "vote": {
+      "description": "Users can vote up to rank the popular queries.",
+      "type": "integer",
+      "default": 1
+    },
+    "query": {
+      "description": "SQL Query definition.",
+      "type": "string"
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:78]: Use basic.json sqlQuery type

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+      "type": "array",
+      "items": {
+        "$ref": "../../type/tagLabel.json"
+      },
+      "default": null
+    },
+    "deleted": {
+      "description": "When `true` indicates the entity has been soft deleted.",
+      "type": "boolean",
+      "default": false
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:99]: @harshach do we need soft delete for queries?

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        dao,
+        QUERY_PATCH_FIELDS,
+        QUERY_UPDATE_FIELDS);
+  }
+
+  @Override
+  public Query setFields(Query entity, EntityUtil.Fields fields) throws IOException {
+    entity.setOwner(fields.contains(FIELD_OWNER) ? getOwner(entity) : null);
+    entity.setTags(fields.contains(FIELD_TAGS) ? getTags(entity.getFullyQualifiedName()) : null);
+    entity.setFollowers(fields.contains(FIELD_FOLLOWERS) ? getFollowers(entity) : null);
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java:46]: Owner, tags, followers etc. are already set by EntityRepository.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    applyTags(queryEntity);
+  }
+
+  @Override
+  public EntityUpdater getUpdater(Query original, Query updated, Operation operation) {
+    return new QueryUpdater(original, updated, operation);
+  }
+
+  public ResultList<Query> listQueriesByEntityId(
+      String entityId, String entityType, String before, String after, int limit) {
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java:134]: Can't we use ListFilter based conditions and reuse the paginated listing?

File: openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json [hunk 5]
```diff
@@ -57,6 +57,16 @@
...
     ]
   },
+  {
+    "name" : "query",
+    "operations" : [
+      "Create",
+      "Delete",
+      "ViewAll",
+      "ViewQueries",
+      "EditAll"
```
Comment (sureshms) [inline openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json:67]: Will queries have description, etc? In which case Edit operations will need to be granular.

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+    "duration": {
+      "description": "How long did the query took to run in seconds.",
+      "type": "number"
+    },
+    "users": {
+      "description": "List of users who ran this query.",
+      "type": "array",
+      "items": {
+        "$ref": "../../type/entityReference.json"
+      },
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:50]: Instead of entityReference, we can just turn this into the names of users. We stopped using EntityReference in create requests.

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+      "type": "string"
+    },
+    "description": {
+      "description": "Description of the query instance.",
+      "$ref": "../../type/basic.json#/definitions/markdown"
+    },
+    "owner": {
+      "description": "Owner of this entity",
+      "$ref": "../../type/entityReference.json",
+      "default": null
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:27]: Can this be both a team or a user? If so entityReference is okay.

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+      },
+      "default": null
+    },
+    "query": {
+      "description": "SQL Query definition.",
+      "type": "string"
+    },
+    "duration": {
+      "description": "How long did the query took to run in seconds.",
+      "type": "number"
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:43]: Use basic.json duration type

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+      "items": {
+        "$ref": "../../type/entityReference.json"
+      },
+      "default": null
+    },
+    "vote": {
+      "description": "Users can vote up to rank the popular queries.",
+      "type": "integer",
+      "default": 1
+    },
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:57]: Vote need not be part of createRequest. Voting happens afterwards once the query is created.

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+import org.openmetadata.service.jdbi3.ListFilter;
+import org.openmetadata.service.jdbi3.QueryRepository;
+import org.openmetadata.service.resources.Collection;
+import org.openmetadata.service.resources.EntityResource;
+import org.openmetadata.service.security.Authorizer;
+import org.openmetadata.service.security.policyevaluator.OperationContext;
+import org.openmetadata.service.util.ResultList;
+
+@Path("/v1/query")
+@Api(value = "Query collection", tags = "query collection")
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:54]: Always rest endpiont should be plural. Change it to `queries`.

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+              schema = @Schema(implementation = Include.class))
+          @QueryParam("include")
+          @DefaultValue("non-deleted")
+          Include include)
+      throws IOException {
+    return getInternal(uriInfo, securityContext, id, fieldsParam, include);
+  }
+
+  @GET
+  @Path("/name/{fqn}")
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:167]: What is fqn for a query? We should delete get-by-name methods if it is not applicable.

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+                        @ExampleObject("[" + "{op:remove, path:/a}," + "{op:add, path: /b, value: val}" + "]")
+                      }))
+          JsonPatch patch)
+      throws IOException {
+    return patchInternal(uriInfo, securityContext, id, patch);
+  }
+
+  @PUT
+  @Path("/{id}/followers")
+  @Operation(
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:350]: @harshach do we want followers for query?

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+          @DefaultValue("false")
+          boolean hardDelete,
+      @Parameter(description = "Id of the query", schema = @Schema(type = "UUID")) @PathParam("id") UUID id)
+      throws IOException {
+    return delete(uriInfo, securityContext, id, false, hardDelete);
+  }
+
+  @PUT
+  @Path("/restore")
+  @Operation(
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:420]: @harshach do we want soft-delete and restore for queries?

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+      })
+  public Response restoreTable(
+      @Context UriInfo uriInfo, @Context SecurityContext securityContext, @Valid RestoreEntity restore)
+      throws IOException {
+    return restoreEntity(uriInfo, securityContext, restore.getId());
+  }
+
+  @PUT
+  @Path("/{id}/addQueryUsage")
+  @Operation(
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:439]: Please avoid adding a verb in the rest API paths. What is addQueryUsage mean? Need better description that explains this API.

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+      @Valid List<EntityReference> entityIds)
+      throws IOException {
+    OperationContext operationContext = new OperationContext(entityType, MetadataOperation.EDIT_ALL);
+    authorizer.authorize(securityContext, operationContext, getResourceContextById(id));
+    return dao.addQueryUsage(id, entityIds);
+  }
+
+  @PUT
+  @Path("/{id}/removeQueryUsage")
+  @Operation(
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:463]: Again this should be DELETE .../queries/{id}/usage. Why is DELETE required?

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+        @ApiResponse(responseCode = "200", description = "OK"),
+        @ApiResponse(responseCode = "404", description = "Query for instance {fqn} is not found")
+      })
+  public Response delete(
+      @Context UriInfo uriInfo,
+      @Context SecurityContext securityContext,
+      @Parameter(description = "Hard delete the entity. (Default = `false`)")
+          @QueryParam("hardDelete")
+          @DefaultValue("false")
+          boolean hardDelete,
```
Comment (sureshms) [inline openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:502]: Questionable if we need soft delete.

File: openmetadata-service/src/test/java/org/openmetadata/service/resources/EntityResourceTest.java [hunk 17]
```diff
@@ -810,7 +812,7 @@ void get_entityIncludeDeleted_200(TestInfo test) throws IOException {
   ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
   @Test
   @Execution(ExecutionMode.CONCURRENT)
-  protected void post_entityCreateWithInvalidName_400() {
+  protected void post_entityCreateWithInvalidName_400() throws IOException {
```
Comment (sureshms) [inline openmetadata-service/src/test/java/org/openmetadata/service/resources/EntityResourceTest.java:815]: Do sneakyThrows at the place where this method is overridden instead of adding this to the base method inherited by everyone.

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json (hunk 9 above)
Comment (harshach) [reply openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:57]: this should be an api, similar to follow/unfollow kind of api

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json (hunk 2 above)
Comment (harshach) [reply openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:99]: I don't think so. we can avoid this.

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    },
+    "queryUsedIn": {
+      "description": "Entities that are using this query",
+      "$ref": "../../type/entityReference.json#/definitions/entityReferenceList"
```
[the remark this answers is not in the document]
Comment (mohityadav766) [reply openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:103]: yup when usage runs and the query is processed query can have joins, and can be associated with multiple table , this fields adds the usage of this query across different entities

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json (hunk 2 above)
Comment (mohityadav766) [reply openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:99]: removed

File: openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json (hunk 5 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json:67]: yes it will have Descriptions

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java (hunk 12 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:350]: Queries can be added through Ingestion, and maybe through the UI as well in the near future, this will allow users who want to follow a frequently used query or a query that have a use for them , let me know if doesn't make sense we can remove it

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java [hunk 19]
```diff
@@ -594,6 +598,21 @@ public SQLQuery map(ResultSet rs, StatementContext ctx) throws SQLException {
     }
   }
 
+  class QueryMapper implements RowMapper<QueryList> {
+    @Override
+    public QueryList map(ResultSet rs, StatementContext ctx) throws SQLException {
+      String fqn = rs.getString("name");
```
[the remark this answers is not in the document]
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java:604]: yup , and name is essentially the query checksum

File: bootstrap/sql/com.mysql.cj.jdbc.Driver/v008__create_db_connection_info.sql [hunk 20]
```diff
@@ -45,7 +45,7 @@ SET json = JSON_INSERT(
 )
 WHERE name = 'columnValuesToBeBetween';
 
-UPDATE pipeline_entity 
+UPDATE pipeline_entity
 SET json = JSON_INSERT(
```
Comment (sureshms) [inline bootstrap/sql/com.mysql.cj.jdbc.Driver/v008__create_db_connection_info.sql:49]: Why is this change required? Note we can never change previously release SQL files. Otherwise Flyway bootstrap will have problems.

File: bootstrap/sql/org.postgresql.Driver/v008__create_db_connection_info.sql [hunk 21]
```diff
@@ -66,5 +66,4 @@ jsonb_build_object('connection',jsonb_build_object(
 'password',json #>'{connection,config,password}',
 'provider',json #>'{connection,config,provider}'
 )), true)
-where servicetype = 'Superset';
-
+where servicetype = 'Superset';
```
Comment (sureshms) [inline bootstrap/sql/org.postgresql.Driver/v008__create_db_connection_info.sql:69]: Again, this change needs to be undone. Otherwise we will have upgrade failures.

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json (hunk 6 above)
Comment (mohityadav766) [reply openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json]: we can have teams as well creating queries

File: openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json (hunk 7 above)
Comment (mohityadav766) [reply openmetadata-spec/src/main/resources/json/schema/api/data/createQuery.json:27]: yes  we can have the owners for the SQL queries created through UI as users or teams

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java (hunk 12 above)
Comment (sureshms) [reply openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:327]: We can make following a separate step. Let's remove this for the time being (once you add an API we can't remove it, especially post 1.0).

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java (hunk 10 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java]: updated

File: openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java (hunk 11 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/resources/query/QueryResource.java:174]: fqn is evaluated as the name -> when created from UI , or the query checksum if the name is not provided

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+      queryUsers.add(new EntityReference().withId(record.getId()).withType(record.getType()));
+    }
+    return queryUsers;
+  }
+
+  @Override
+  public void prepare(Query entity) throws IOException {
+    try {
+      if (CommonUtil.nullOrEmpty(entity.getName())) {
+        byte[] checksum = MessageDigest.getInstance("MD5").digest(entity.getQuery().getBytes());
```
[the remark this answers is not in the document]
Comment (mohityadav766) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java]: done

File: openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json [hunk 23]
```diff
@@ -57,6 +57,20 @@
...
     ]
   },
+  {
+    "name" : "query",
+    "operations" : [
+      "Create",
+      "Delete",
+      "ViewAll",
+      "ViewQueries",
+      "EditAll",
```
Comment (sureshms) [inline openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json:67]: Do we need ViewQueries given the entity is a Query entity?

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+        "$ref": "../../type/tagLabel.json"
+      },
+      "default": null
+    },
+    "queryUsedIn": {
+      "description": "Entities that are using this query",
+      "$ref": "../../type/entityReference.json#/definitions/entityReferenceList"
+    },
+    "deleted": {
+      "description": "When `true` indicates the entity has been soft deleted.",
```
Comment (sureshms) [inline openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:101]: If soft delete is not supported, we can remove this deleted flag.

**Review by sureshms — CHANGES_REQUESTED** [review]

File: openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json (hunk 23 above)
Comment (mohityadav766) [reply openmetadata-service/src/main/resources/json/data/ResourceDescriptors.json:67]: As mentioned, if we ever want to be able to create Queries from UI then this will be needed, otherwise, if we are only aiming for ingestion to put Queries then I can change as per that, let me know!

File: bootstrap/sql/org.postgresql.Driver/v008__create_db_connection_info.sql (hunk 21 above)
Comment (mohityadav766) [reply bootstrap/sql/org.postgresql.Driver/v008__create_db_connection_info.sql]: my IntelliJ, removed the change

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json (hunk 24 above)
Comment (mohityadav766) [reply openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:101]: removed

**Review by harshach — APPROVED** [review]

**Review by sureshms — APPROVED** [review]
